File: docs/palindromes-in-peg/dp_search_reuse.py
"""Cancellation/reentry for the whole offline doubling search.

One extra unary tape tracks WINDOW's distance from C and furthest visit.
Finite cancellation entries finish a half-completed paired move before
cleanup. Cancel any normal instruction, including internal kernel cleanup,
but not the outer cleanup itself. WINDOW/LOWER survive; all scratch clears.
"""
import copy

from fpp_finite import Program, LEFT, END, BLANK, C
from fpp_reuse import C_ORIGIN
from dp_finite import LOWER
from dp_search_finite import WINDOW, STATUS, period_symbol

ERASED = "erased:_"


def make_cancellable(search):
  if search.ntapes < 16:
    raise ValueError("expected the local doubling search or its chain monitor")
  distance = search.ntapes
  p = Program(search.alphabet, ntapes=search.ntapes + 1)
  p.distance = distance
  p.scalar_tapes = tuple(getattr(search, "scalar_tapes", (STATUS,)))
  p.source_alphabet = search.source_alphabet
  p.code = copy.deepcopy(search.code)
  p.found, p.missed = search.found, search.missed
  for name in ("outcomes", "ready", "confirmed_ready"):
    if hasattr(search, name):
      setattr(p, name, copy.deepcopy(getattr(search, name)))
  p.scratch = tuple(t for t in range(p.ntapes) if t not in (WINDOW, LOWER))

  # Retain origins during INTERNAL cleanup so a cancellation can locate
  # them even halfway through that sweep. The next kernel bootstrap
  # overwrites these same origin cells; its body still starts blank.
  for row in p.code:
    if row[0] != "read":
      continue
    tape = row[1]
    root = C_ORIGIN if tape == C else LEFT
    target = row[2].get(root)
    if target is not None and tuple(p.code[target][:3]) == ("write", tape, BLANK):
      p.code[target] = ("write", tape, root, p.code[target][3])

  # Logical deletion leaves a physical tombstone. Internal left-to-right
  # cleanup can otherwise leave a blank hole before still-live cells.
  # Normal code treats the tombstone exactly as blank; outer cleanup sees
  # a dense physical prefix and sweeps through it to the true frontier.
  for q, row in enumerate(p.code):
    if row[0] == "write" and row[2] == BLANK:
      p.code[q] = ("write", row[1], ERASED, row[3])
    elif row[0] == "read" and BLANK in row[2]:
      p.code[q] = ("read", row[1], {**row[2], ERASED: row[2][BLANK]})

  pending = {}
  for q, row in enumerate(search.code):
    if row[0] == "move" and row[1] == WINDOW:
      direction, k = row[2:]
      if direction == -1:
        written = p.write(distance, "1", k)
        pending[written] = "write"
        k = written
      moved = p.move(distance, -direction, k)
      pending[moved] = -direction
      p.code[q] = ("move", WINDOW, direction, moved)

  done = p.add(("halt",))
  symbols = [set() for _ in range(p.ntapes)]
  for row in p.code:
    if row[0] == "read":
      symbols[row[1]].update(row[2])
    elif row[0] == "write":
      symbols[row[1]].add(row[2])

  def reset(tape, k):
    root = C_ORIGIN if tape == C else LEFT
    body = sorted((symbols[tape] | {BLANK}) - {root})
    seek, clear, back = p.reserve(), p.reserve(), p.reserve()
    p.branch(tape, {root: p.write(tape, BLANK, k),
                    BLANK: p.move(tape, -1, back)}, back)
    p.branch(tape, {BLANK: p.move(tape, -1, back), **{
      s: p.write(tape, BLANK, p.move(tape, 1, clear))
      for s in body if s != BLANK}}, clear)
    p.branch(tape, {root: p.move(tape, 1, clear), **{
      s: p.move(tape, -1, seek) for s in body}}, seek)
    return seek

  clean_scratch = done
  for tape in reversed(p.scratch):
    if tape in p.scalar_tapes:
      clean_scratch = p.write(tape, BLANK, clean_scratch)
    else:
      clean_scratch = reset(tape, clean_scratch)
  lower = p.reserve()
  p.branch(LOWER, {LEFT: clean_scratch, **{
    s: p.move(LOWER, -1, lower) for s in ("1", END)}}, lower)

  # First return to C using the aligned distance heads. Sweep precisely
  # the already visited prefix, erasing P tags, then return to C again.
  # The retained unary extent makes marks behind the current head visible.
  home, scan, restore = p.reserve(), p.reserve(), p.reserve()
  p.branch(distance, {LEFT: lower,
    "1": p.move(WINDOW, 1, p.move(distance, -1, restore))}, restore)
  next_cell = p.move(distance, 1, scan)
  unmark = p.branch(WINDOW, {LEFT: next_cell, **{
    s: next_cell for s in p.source_alphabet}, **{
    period_symbol(s): p.write(WINDOW, s, next_cell) for s in p.source_alphabet}})
  p.branch(distance, {BLANK: p.move(distance, -1, restore),
    "1": p.move(WINDOW, -1, unmark)}, scan)
  p.branch(distance, {LEFT: next_cell,
    "1": p.move(WINDOW, 1, p.move(distance, -1, home))}, home)
  p.cleanup = home

  p.cancel_entries = {q: home for q in range(len(search.code))}
  for q, action in pending.items():
    if action == "write":
      entry = p.write(distance, "1", home)
    elif action == 1:
      entry = p.move(distance, 1, p.write(distance, "1", home))
    else:
      entry = p.move(distance, -1, home)
    p.cancel_entries[q] = entry

  def bootstrap(k):
    states = []
    for tape in reversed(p.scratch):
      symbol = ("N" if tape == STATUS else BLANK if tape in p.scalar_tapes
                else C_ORIGIN if tape == C else LEFT)
      k = p.write(tape, symbol, k)
      states.append(k)
    return k, states

  cancel_bootstrap, _ = bootstrap(home)
  p.start, states = bootstrap(search.start)
  p.cancel_entries.update({q: cancel_bootstrap for q in states})
  p.validate()
  return p

File: scala/pal/src/main/scala/pal/DpSearchReuse.scala
package pal

import scala.collection.mutable

import FppFinite.*
import FppFinite.Instruction.*
import FppReuse.C_ORIGIN
import DpFinite.LOWER
import DpSearchFinite.{WINDOW, STATUS, periodSymbol}

/** Cancellation/reentry for the whole offline doubling search.
  *
  * One extra unary tape tracks WINDOW's distance from C and furthest visit.
  * Finite cancellation entries finish a half-completed paired move before
  * cleanup. Cancel any normal instruction, including internal kernel cleanup,
  * but not the outer cleanup itself. WINDOW/LOWER survive; all scratch clears.
  *
  * Python 原典: `dp_search_reuse.py`。
  */
object DpSearchReuse {
  val ERASED: String = "erased:_"

  /** What a paired WINDOW/DISTANCE move had left undone when cancelled at that state. */
  sealed trait Pending
  object Pending {
    /** The unary distance mark write is still due. */
    case object Write extends Pending
    /** The DISTANCE head still has to move by `direction`. */
    final case class Move(direction: Int) extends Pending
  }

  /** Python の `make_cancellable(search)`。 */
  def makeCancellable(search: Program): Program = {
    if (search.ntapes < 16) {
      throw new IllegalArgumentException("expected the local doubling search or its chain monitor")
    }
    val distance = search.ntapes
    val p = new Program(search.alphabet, ntapes = search.ntapes + 1)
    p.distance = Some(distance)
    p.scalarTapes = Some(search.scalarTapes.getOrElse(Vector(STATUS)))
    val scalarTapes = p.scalarTapes.get
    p.sourceAlphabet = search.sourceAlphabet
    p.copyCodeFrom(search)
    p.found = search.found
    p.missed = search.missed
    p.outcomes ++= search.outcomes
    p.ready = search.ready
    p.confirmedReady = search.confirmedReady
    p.scratch = (0 until p.ntapes).filterNot(t => t == WINDOW || t == LOWER).toVector

    retainOriginsDuringInternalCleanup(p)
    replaceBlankWritesByTombstones(p)

    val pending = mutable.LinkedHashMap.empty[Int, Pending]
    for ((row, q) <- search.instructions.zipWithIndex) {
      row match {
        case Move(WINDOW, direction, next) =>
          var k = next
          if (direction == -1) {
            val written = p.write(distance, "1", k)
            pending(written) = Pending.Write
            k = written
          }
          val moved = p.move(distance, -direction, k)
          pending(moved) = Pending.Move(-direction)
          p.set(q, Move(WINDOW, direction, moved))
        case _ => ()
      }
    }

    val done = p.add(Halt)
    val symbols = symbolsPerTape(p)

    def reset(tape: Int, k: Int): Int = {
      val root = if (tape == C) { C_ORIGIN } else { LEFT }
      val body = ((symbols(tape) + BLANK) - root).toVector.sorted
      val seek = p.reserve()
      val clear = p.reserve()
      val back = p.reserve()
      p.branch(tape, choices(root -> p.write(tape, BLANK, k),
        BLANK -> p.move(tape, -1, back)), back)
      p.branch(tape, choices((Seq(BLANK -> p.move(tape, -1, back)) ++ body.filter(_ != BLANK).map { s =>
        s -> p.write(tape, BLANK, p.move(tape, 1, clear))
      })*), clear)
      p.branch(tape, choices((Seq(root -> p.move(tape, 1, clear)) ++ body.map { s =>
        s -> p.move(tape, -1, seek)
      })*), seek)
      seek
    }

    var cleanScratch = done
    for (tape <- p.scratch.reverse) {
      if (scalarTapes.contains(tape)) {
        cleanScratch = p.write(tape, BLANK, cleanScratch)
      } else {
        cleanScratch = reset(tape, cleanScratch)
      }
    }
    val lower = p.reserve()
    p.branch(LOWER, choices((Seq(LEFT -> cleanScratch) ++ Seq("1", END).map { s =>
      s -> p.move(LOWER, -1, lower)
    })*), lower)

    // First return to C using the aligned distance heads. Sweep precisely
    // the already visited prefix, erasing P tags, then return to C again.
    // The retained unary extent makes marks behind the current head visible.
    val home = p.reserve()
    val scan = p.reserve()
    val restore = p.reserve()
    p.branch(distance, choices(LEFT -> lower,
      "1" -> p.move(WINDOW, 1, p.move(distance, -1, restore))), restore)
    val nextCell = p.move(distance, 1, scan)
    val unmark = p.branch(WINDOW, choices((Seq(LEFT -> nextCell) ++
      p.sourceAlphabet.map(s => s -> nextCell) ++
      p.sourceAlphabet.map(s => periodSymbol(s) -> p.write(WINDOW, s, nextCell)))*))
    p.branch(distance, choices(BLANK -> p.move(distance, -1, restore),
      "1" -> p.move(WINDOW, -1, unmark)), scan)
    p.branch(distance, choices(LEFT -> nextCell,
      "1" -> p.move(WINDOW, 1, p.move(distance, -1, home))), home)
    p.cleanup = Some(home)

    for (q <- search.code.indices) {
      p.cancelEntries(q) = home
    }
    for ((q, action) <- pending) {
      val entry = action match {
        case Pending.Write => p.write(distance, "1", home)
        case Pending.Move(1) => p.move(distance, 1, p.write(distance, "1", home))
        case Pending.Move(_) => p.move(distance, -1, home)
      }
      p.cancelEntries(q) = entry
    }

    def bootstrap(k: Int): (Int, Vector[Int]) = {
      val states = Vector.newBuilder[Int]
      var next = k
      for (tape <- p.scratch.reverse) {
        val symbol =
          if (tape == STATUS) { "N" }
          else if (scalarTapes.contains(tape)) { BLANK }
          else if (tape == C) { C_ORIGIN }
          else { LEFT }
        next = p.write(tape, symbol, next)
        states += next
      }
      (next, states.result())
    }

    val (cancelBootstrap, _) = bootstrap(home)
    val (start, states) = bootstrap(search.start)
    p.start = start
    for (q <- states) {
      p.cancelEntries(q) = cancelBootstrap
    }
    p.validate()
    p
  }

  /** Retain origins during INTERNAL cleanup so a cancellation can locate
    * them even halfway through that sweep. The next kernel bootstrap
    * overwrites these same origin cells; its body still starts blank.
    */
  private def retainOriginsDuringInternalCleanup(p: Program): Unit = {
    for (q <- p.code.indices) {
      p.instruction(q) match {
        case Read(tape, choices) =>
          val root = if (tape == C) { C_ORIGIN } else { LEFT }
          for (target <- choices.get(root)) {
            p.instruction(target) match {
              case Write(`tape`, BLANK, next) => p.set(target, Write(tape, root, next))
              case _ => ()
            }
          }
        case _ => ()
      }
    }
  }

  /** Logical deletion leaves a physical tombstone. Internal left-to-right
    * cleanup can otherwise leave a blank hole before still-live cells.
    * Normal code treats the tombstone exactly as blank; outer cleanup sees
    * a dense physical prefix and sweeps through it to the true frontier.
    */
  private def replaceBlankWritesByTombstones(p: Program): Unit = {
    for ((row, q) <- p.instructions.zipWithIndex) {
      row match {
        case Write(tape, BLANK, next) => p.set(q, Write(tape, ERASED, next))
        case Read(tape, choices) if choices.contains(BLANK) =>
          p.set(q, Read(tape, choices.updated(ERASED, choices(BLANK))))
        case _ => ()
      }
    }
  }

  /** Every symbol read or written on each tape. */
  private def symbolsPerTape(p: Program): Vector[Set[String]] = {
    val symbols = Vector.fill(p.ntapes)(mutable.HashSet.empty[String])
    for (row <- p.instructions) {
      row match {
        case Read(tape, choices) => symbols(tape) ++= choices.keys
        case Write(tape, symbol, _) => symbols(tape) += symbol
        case _ => ()
      }
    }
    symbols.map(_.toSet)
  }
}
